File: modules/expo-bluesky-video-player/ios/ExpoBlueskyVideoPlayerModule.swift
import ExpoModulesCore

public class ExpoBlueskyVideoPlayerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBlueskyVideoPlayer")
    
    AsyncFunction("prefetchAsync") { (source: String, promise: Promise) in
      PlayerItemManager.shared.saveToCache(source: source)
      promise.resolve()
    }

    View(ExpoBlueskyVideoPlayerView.self) {
      Events(
        "onPlayerStateChange"
      )
      
      Prop("source") { (view: ExpoBlueskyVideoPlayerView, prop: String) in
        view.source = prop
      }
      
      Prop("autoplay") { (view: ExpoBlueskyVideoPlayerView, prop: Bool) in
        view.autoplay = prop
      }
      
      AsyncFunction("toggleAsync") { (view: ExpoBlueskyVideoPlayerView) in
        view.toggle()
      }
      
      AsyncFunction("playAsync") { (view: ExpoBlueskyVideoPlayerView) in
        view.play()
      }
      
      AsyncFunction("pauseAsync") { (view: ExpoBlueskyVideoPlayerView) in
        view.pause()
      }
    }
  }
}

File: modules/expo-bluesky-video-player/android/src/main/java/expo/modules/blueskyvideoplayer/ExpoBlueskyVideoPlayerModule.kt
package expo.modules.blueskyvideoplayer

import androidx.media3.common.util.UnstableApi
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

@UnstableApi
class ExpoBlueskyVideoPlayerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBlueskyVideoPlayer")

    AsyncFunction("prefetchAsync") {
      { source: String, promise: Promise ->
        MediaItemManager(appContext).saveToCache(source)
        promise.resolve()
      }
    }

    View(ExpoBlueskyVideoPlayerView::class) {
      Events(
        "onPlayerStateChange"
      )

      Prop("source") { view: ExpoBlueskyVideoPlayerView, source: String ->
          view.source = source
      }

      Prop("autoplay") { view: ExpoBlueskyVideoPlayerView, autoplay: Boolean ->
        view.autoplay = autoplay
      }

      AsyncFunction("playAsync") { view: ExpoBlueskyVideoPlayerView ->
        view.play()
      }

      AsyncFunction("pauseAsync") { view: ExpoBlueskyVideoPlayerView ->
        view.pause()
      }

      AsyncFunction("toggleAsync") { view: ExpoBlueskyVideoPlayerView ->
        view.toggle()
      }
    }
  }
}
